### ISSUE TITLE
Image files written by s4vol unusable

### ISSUE THREAD

**Issue by AlainKnaff** [dgesswein/s4-3b1-pc7300#6]
Hi,

While trying to use s4vol to create disk images with bigger paging (swap) space, I came across several issues:
1. In parameter parsing, the "-io" option lacked a continue => option cannot be used, complains that file name is missing even if it is indeed given.
2. When specifying a page partition size, number of tracks for page was not recalculated, leading to a page partition that had the minimum size
3. While writing out the new image, there was a miscalculation of the offset of the bad block table (division instead of multiplication by sector size) => such images can't even be parsed by s4disk!
4. Checksum of disk label was left zero, hoping that a zero value would just be interpreted as "don't verify". But unfortunately this does not work in the emulator, which fails to find the disk (or the UNIX3.51 file) during boot. Not sure how a physical UnixPC would behave here.

Attached is a patch to fix these items.
[s4vol-patch.txt](https://github.com/dgesswein/s4-3b1-pc7300/files/9847226/s4vol-patch.txt)

Thanks,

Alain

**Comment by arnoldrobbins** [comment]
@AlainKnaff This looks interesting. I think (or thought) there was a limit to the size of the swap partition in the 3B1 kernel. Do you know what it is? Have you managed to make swap space bigger than that limit? Thanks.

**Comment by AlainKnaff** [comment]
Hi Arnold,

On 23/10/2022 21:16, Arnold Robbins wrote:
> @AlainKnaff This looks interesting. I think (or thought) there was a limit to the size of the swap partition in the 3B1 kernel. Do you know what it is?

No idea, but in any case, at 40M it is not yet reached :-)

There is however a limit on the size of the virtual memory for *one*
process, which is 3 Megabytes. But if many big processes coexist, huge
swap is useful.


> Have you managed to make swap space bigger than that limit? Thanks.
> 

I tried 40M. I didn't try to "max it out" yet.

Regards,

Alain

**Comment by AlainKnaff** [comment]
Hi all,

On 23/10/2022 21:36, Alain Knaff wrote:
[...]
> I tried 40M. I didn't try to "max it out" yet.

So, I just did now.

After about 111 Megs, the UnixPC (emulator) no longer boots, but maybe
because the huge swap pushes the data partition (which contains the
kernel) too far out of the way.

Until 111 Megs, it boots ok.

Unfortunately, I don't know the equivalent of the free command on
UnixPC, so I had to resort to an experimental approach to see how much
memory it actually had available. I wrote a small program that attempts
to malloc() as much memory as it can get, and then backgrounds itself.
Moreover, this program takes care to write to each page of memory that
it got.

Each instance manages to allocate 2426 KBytes (but probably consumes
somewhat more memory for misc housekeeping, etc.)

I can spawn 22 of these before I get "fork: No more processes", that
would be 53372 KBytes. So I presume that there is a limit of 64 Megs
beyond which additional swap is no longer used.

Regards,

Alain

**Comment by dgesswein** [comment]
Thanks for the fix. Committed.

**Comment by dgesswein** [comment]
I saw your Usenet posting originally asking about why swap space on your disks wasn't being used. Didn't know the answer. Did you figure out why it wasn't using your swap initially and if it can use swap on multiple disks?

**Comment by AlainKnaff** [comment]
Hi,

On 24/10/2022 03:33, dgesswein wrote:
> I saw your Usenet posting originally asking about why swap space on your disks wasn't being used. Didn't know the answer.

No problem

> Did you figure out why it wasn't using your swap initially

Actually, my swap from the first disk (4M) *was* being used. I had just
miscalculated memory consumption.

> and if it can use swap on multiple disks?
> 

No, unfortunately swap partitions from a second disk are not picked up,
I just tried it again.

Back in May, I eventually resorted to re-install from floppies and chose
a bigger swap size during set up. I then added Arnold's "full disk with
lots of software" as a second disk and chrooted into it.

Regards,

Alain
